hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclController.java

```diff
@@ -0,0 +1,606 @@
+    if (hdfsAclHelper != null) {
+      hdfsAclHelper.snapshotAcl(snapshot);
+    }
+  }
+
+  @Override
+  public void postCompletedTruncateTableAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      TableName tableName) throws IOException {
+    if (hdfsAclHelper != null && !tableName.isSystemTable()) {
+      hdfsAclHelper.resetTableAcl(tableName);
+    }
+  }
+
+  @Override
+  public void postDeleteTable(ObserverContext<MasterCoprocessorEnvironment> ctx,
+      TableName tableName) throws IOException {
+    if (hdfsAclHelper != null && !tableName.isSystemTable()) {
+      // remove table user HDFS acl from ns data directory if the user has no global/the ns/other
+      // tables of the ns permissions
+      Set<String> removeUsers = new HashSet<>();
+      try (Table aclTable =
+          ctx.getEnvironment().getConnection().getTable(PermissionStorage.ACL_TABLE_NAME)) {
+        List<String> users = SnapshotScannerHDFSAclStorage.getTableUsers(aclTable, tableName);
+        SnapshotScannerHDFSAclStorage.deleteTableHdfsAcl(aclTable, tableName);
+        for (String user : users) {
+          List<byte[]> userEntries = SnapshotScannerHDFSAclStorage.getUserEntries(aclTable, user);
+          boolean remove = true;
+          for (byte[] entry : userEntries) {
+            if (PermissionStorage.isGlobalEntry(entry)) {
+              remove = false;
+              break;
+            } else if (PermissionStorage.isNamespaceEntry(entry) && Bytes
+                .equals(PermissionStorage.fromNamespaceEntry(entry), tableName.getNamespace())) {
+              remove = false;
+              break;
+            } else if (Bytes.equals(TableName.valueOf(entry).getNamespace(),
+              tableName.getNamespace())) {
+              remove = false;
+              break;
+            }
+          }
+          if (remove) {
+            removeUsers.add(user);
+          }
+        }
+      }
+      if (removeUsers.size() > 0) {
+        hdfsAclHelper.removeNamespaceAcl(tableName, removeUsers);
+      }
+    }
+  }
+
+  @Override
+  public void postDeleteNamespace(ObserverContext<MasterCoprocessorEnvironment> ctx,
+      String namespace) throws IOException {
+    if (hdfsAclHelper != null) {
+      try (Table aclTable =
+          ctx.getEnvironment().getConnection().getTable(PermissionStorage.ACL_TABLE_NAME)) {
+        SnapshotScannerHDFSAclStorage.deleteNamespaceHdfsAcl(aclTable, namespace);
+      }
+      Path tmpNsDir = pathHelper.getTmpNsDir(namespace);
+      if (fs.exists(tmpNsDir)) {
+        if (fs.listStatus(tmpNsDir).length == 0) {
+          fs.delete(tmpNsDir, true);
+        } else {
+          LOG.error("The tmp directory of namespace {} is not empty after delete namespace",
+            namespace);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void postGrant(ObserverContext<MasterCoprocessorEnvironment> c,
+      UserPermission userPermission, boolean mergeExistingPermissions) throws IOException {
```

Comment:
Need to handle the case when mergeExistingPermissions is true?

Reply:
It's considered because we get the merged permission from acl table and check if a user has read permission based on the merged permission.